### PR TITLE
POC: MapReduce-like scheduler priority functions

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #24246

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31606)

### CODE REVIEW

File: plugin/pkg/scheduler/algorithm/priorities/priorities_test.go [hunk 1]
```diff
@@ -52,7 +55,26 @@ func makeNode(node string, milliCPU, memory int64) *api.Node {
...
+		if reduceFn != nil {
+			if err := reduceFn(result); err != nil {
+				return nil, err
+			}
+		}
+		return result, nil
+	}
+}
+
+/*func TestZeroRequest(t *testing.T) {
```
Comment (wojtek-t) [inline plugin/pkg/scheduler/algorithm/priorities/priorities_test.go]: I think that fixing this test, is the only missing part of this PR for now.

In general, changing all priority functions to the new signature may require some small changes, but I took a look into e.g. SelectorSpreading (which seems to be the most problematic), and it should fit this framework.

**Comment by davidopp** [comment]
This approach/interface looks great to me.

Just curious, why did you move the test from prioities_test.go to generic_scheduler_test.go?

**Comment by wojtek-t** [comment]
>
Just curious, why did you move the test from prioities_test.go to generic_scheduler_test.go?

That test was relying on "plugin/pkg/scheduler", which was causing circular depencies. And in fact, this test was more about scheduler itself than about particular priority functions.

**Comment by wojtek-t** [comment]
@gmarek - can you please review this PR (it's not super urgent, since it won't be merged before branch will be opened for 1.5 anyway).

**Comment by davidopp** [comment]
I actually reviewed the whole PR as it currently is; I don't think @gmarek needs to review it if you are planning to submit this. If you are planning to do more work, then I guess we can decide once that work is done.

**Comment by wojtek-t** [comment]
@davidopp - thanks, I wasn't planning to extend this PR - I think it would be easier to do further changes in separate PR/PRs.
So if you 're fine with it, can you please apply lgtm label (then it will automatically merge when we open repo for 1.5 changes).

**Comment by davidopp** [comment]
LGTM

File: plugin/pkg/scheduler/algorithm/types.go [hunk 2]
```diff
@@ -26,9 +26,24 @@ import (
 // The failure information is given by the error.
 type FitPredicate func(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (bool, []PredicateFailureReason, error)
 
+// PriorityMapFunction is a function that computes per-node results for a given node.
+// TODO: Figure out the exact API of this method.
+type PriorityMapFunction func(pod *api.Pod, meta interface{}, nodeInfo *schedulercache.NodeInfo) (schedulerapi.HostPriority, error)
```
Comment (gmarek) [inline plugin/pkg/scheduler/algorithm/types.go]: I don't like the `meta interface{}` part...

File: plugin/pkg/scheduler/generic_scheduler.go [hunk 3]
```diff
@@ -262,8 +264,26 @@ func PrioritizeNodes(
...
-			priorityFunc := config.Function
-			prioritizedList, err := priorityFunc(pod, nodeNameToInfo, nodes)
+
+			prioritizedList, err := func() (schedulerapi.HostPriorityList, error) {
+				if config.Function != nil {
+					return config.Function(pod, nodeNameToInfo, nodes)
+				}
+				prioritizedList := make(schedulerapi.HostPriorityList, 0, len(nodes))
+				for i := range nodes {
+					hostResult, err := config.Map(pod, meta, nodeNameToInfo[nodes[i].Name])
```
Comment (gmarek) [inline plugin/pkg/scheduler/generic_scheduler.go]: Shouldn't it run in yet another go-routine?

Comment (wojtek-t) [inline plugin/pkg/scheduler/generic_scheduler.go]: At least not initially. Once everything is using this pattern, we will be in better position for optimizations and parallelization at the framework basis.

**Pull request merged**